File: src/lc_0208.rs
// Trie (Prefix Tree)
use std::collections::HashSet;

#[derive(Default)]
pub struct Trie {
    data: HashSet<String>,
}
// this isnt the actual Trie implementation that they expect
// as such this code actually runs super slow! (beats 5%)

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {
    pub fn new() -> Self {
        Trie {
            data: HashSet::new(),
        }
    }

    pub fn insert(&mut self, word: String) {
        self.data.insert(word);
    }

    pub fn search(&self, word: String) -> bool {
        self.data.contains(&word)
    }

    pub fn starts_with(&self, prefix: String) -> bool {
        for k in self.data.iter() {
            if k.starts_with(&prefix) {
                return true;
            }
        }
        false
    }
}

File: leetcode-go/problems/trie/lc_0208.go
package trie

type Trie struct {
	children       [26]*Trie
	isCompleteWord bool
}

func Constructor() Trie {
	return Trie{}
}

func (this *Trie) Insert(word string) {
	node := this
	for _, char := range word {
		idx := char - 'a'
		if node.children[idx] == nil {
			node.children[idx] = &Trie{}
		}
		node = node.children[idx]
	}
	node.isCompleteWord = true
}

func (this *Trie) SearchNode(word string) *Trie {
	node := this
	for _, char := range word {
		idx := char - 'a'
		if child := node.children[idx]; child != nil {
			node = child
		} else {
			return nil
		}
	}
	return node
}

func (this *Trie) Search(word string) bool {
	node := this.SearchNode(word)
	return node != nil && node.isCompleteWord
}

func (this *Trie) StartsWith(prefix string) bool {
	node := this.SearchNode(prefix)
	return node != nil
}
